cw/cw305/ceca.py

```diff
@@ -0,0 +1,574 @@
+    """Signals ``TraceWorker`` instances to filter noisy traces.
+
+    Args:
+        workers:``TraceWorker`` handles.
+        mean_trace: Mean of all traces.
+        std_trace: Standard deviation of all traces.
+        max_std: Allowed number of standard deviations from the mean trace.
+
+    Returns:
+        Number of remaining traces.
+    """
+    min_trace = mean_trace - max_std * std_trace
+    max_trace = mean_trace + max_std * std_trace
+    tasks = [
+        worker.filter_noisy_traces.remote(min_trace, max_trace)
+        for worker in workers
+    ]
+
+    running_cnt = 0
+    while tasks:
+        done, tasks = ray.wait(tasks)
+        running_cnt += ray.get(done[0])
+    return running_cnt
+
+
+def compute_mean_plaintext_traces(workers):
+    """Computes mean traces for all values of all plaintext bytes.
+
+    This function distributes work to ``TraceWorker`` instances and collects
+    their results.
+
+    Args:
+        workers: ``TraceWorker`` handles.
+
+    Returns:
+        A (16, 256, NUM_SAMPLES) array A, where A[i, j, :] is the mean of all
+        traces where plaintext byte i is j.
+    """
+    tasks = [
+        worker.count_and_sum_plaintext_traces.remote() for worker in workers
+    ]
+
+    running_cnt = None
+    running_sum = None
+    while tasks:
+        done, tasks = ray.wait(tasks)
+        cnt, sum_ = ray.get(done[0])
+        if running_cnt is None:
+            running_cnt = np.copy(cnt)
+            running_sum = np.copy(sum_)
+        else:
+            running_cnt += cnt
+            running_sum += sum_
+
+    return running_sum / running_cnt
+
+
+@timer()
+def compute_pairwise_diffs_and_scores(mean_traces):
+    """Computes pairwise differences and their confidence scores between key
+    bytes.
+
+    This function correlates mean plaintext traces to pick the most likely
+    differences.
+
+    Args:
+        mean_traces: A (16, 256, NUM_SAMPLES) array of mean plaintext traces.
+
+    Returns:
+        A (16, 16, 2) array A, where A[i, j, 0] is the difference with the
+        largest correlation coefficient between plaintext bytes i and j, and
+        A[i, j, 1] is the corresponding confidence score.
+    """
+    pairwise_diffs_scores = np.zeros((16, 16, 2))
+    # All possible values of plaintext byte a.
+    alphas = np.arange(256)
+    # All possible differences.
+    diffs = np.arange(256)[:, np.newaxis]
+    # Values of plaintext byte b for all possible differences: A (256, 256)
+    # array where betas[i, j] = diffs[i] ^ alphas[j]
+    betas = alphas ^ diffs
+    # Compute the most likely differences between all pairs of plaintext bytes.
+    for a in range(16):
+        for b in range(a + 1, 16):
+            corrcoefs = np.corrcoef(mean_traces[a], mean_traces[b])
+            # np.corrcoef returns the values that we want in the upper right
+            # quadrant.
+            diff_corrcoefs = corrcoefs[alphas, 256 + betas].sum(axis=1)
+            best_diff = diff_corrcoefs.argmax()
```

Comment:
It's not fully clear how your implementation works but you basically managed to skip three more loops (compared to mine) which is great :+1:

Reply:
What I am wondering: you don't seem to normalize the correlation coefficients BEFORE doing the argmax. If I remember correctly, this was necessary in my implementation. But still yours works correctly....

Reply:
> What I am wondering: you don't seem to normalize the correlation coefficients BEFORE doing the argmax. If I remember correctly, this was necessary in my implementation. But still yours works correctly....

IIUC, you are referring to this snippet (lines [98](https://github.com/lowRISC/ot-sca/blob/8e61b89075ae5a886d35090b2675d721705d495c/cw/cw305/correlation-enhanced_collision_attack.py#L98)-104 in `correlation-enhanced_collision_attack.py`):
```python
    for i in range(256):
        for j in range(256):
            delta = i ^ j
            rho_avg_delta[delta] += rho_mat[i, j]
            rho_avg_delta_num[delta] += 1
    for delta in range(256):                                # redundant
        rho_avg_delta[delta] /= rho_avg_delta_num[delta]    # redundant
    # Normalize correlations.
    mean = rho_avg_delta.mean()
    rho_avg_delta /= mean
```

Unless I'm missing something, it looks like the loop marked by `redundant` doesn't really do anything, maybe except for introducing some tiny numerical error due to floating point operations. Since `rho_avg_delta_num[delta]` is `256` for all values of `delta`, i.e. [0, 255] and `rho_avg_delta` is normalized again, this loop is basically dividing all entries by some constant before they are normalized.

Reply:
> It's not fully clear how your implementation works but you basically managed to skip three more loops (compared to mine) which is great 👍

Thanks :) The main point here is to minimize the back and forth between python and numpy as much as possible, e.g. by avoiding unnecessary (nested) loops that can be handled by numpy much more efficiently. In a nutshell:
- `alphas` is a (256,) row array, `diffs` is a (256, 1) col array (so that broadcasting works out), and `betas` is a (256, 256) array where `betas[i,j] = diffs[i] ^ alphas[j]`.
- Defining `betas` this way is important because this lets us pick the values we want from the `np.corrcoeff` result using just a single indexing operation and a `sum(axis=1)` as opposed to two `range(0, 256)` nested loops.

I hope this helps and please let me know if you have any other questions.

Reply:
Thanks for the clarification @alphan .

Regarding the redundant loop above: I added this (before doing the real normalization) because I wasn't sure if there are equally many combinations of `i` and `j` that result in the same `delta`. But yes, I think with the real normalization in place this loop can actually be dropped.